File: app-android/app/src/main/java/com/aengix/hopper/data/HopperConf.kt
package com.aengix.hopper.data

import com.aengix.hopper.model.HopConstants
import com.aengix.hopper.model.HopNodeProfile
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.PKCS5S2ParametersGenerator
import org.bouncycastle.crypto.params.KeyParameter
import org.json.JSONArray
import org.json.JSONObject
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** Encrypted `.hopperconf` file format + plaintext QR payload kinds. */
object HopperConf {
    const val FILE_EXTENSION = "hopperconf"
    const val MIME_TYPE = "application/x-hopperconf"
    const val ENVELOPE_VERSION = 1
    const val PAYLOAD_VERSION = 1
    const val PBKDF2_ITERATIONS = 210_000
    const val SALT_LENGTH = 16
    const val NONCE_LENGTH = 12
    const val KEY_LENGTH = 32
    const val GCM_TAG_BITS = 128

    object Field {
        const val VERSION = "v"
        const val FMT = "fmt"
        const val FMT_VALUE = "hopperconf"
        const val ALG = "alg"
        const val ALG_VALUE = "aes-256-gcm"
        const val KDF = "kdf"
        const val KDF_VALUE = "pbkdf2-sha256"
        const val ITER = "iter"
        const val SALT = "salt"
        const val NONCE = "nonce"
        const val DATA = "data"
        const val KIND = "kind"
        const val NAME = "name"
        const val SERVER = "server"
        const val HOPS = "hops"
    }

    sealed class Payload {
        data class Server(val profile: HopNodeProfile) : Payload()
        data class Chain(val name: String, val hops: List<HopNodeProfile>) : Payload()
    }

    sealed class ConfError(message: String) : Exception(message) {
        data object Empty : ConfError("The file is empty.")
        data object InvalidEnvelope : ConfError("Not a valid .hopperconf file.")
        data object InvalidPayload : ConfError("The file contents are invalid.")
        data object DecryptionFailed : ConfError("Could not decrypt — check the password.")
        data object EncryptionFailed : ConfError("Could not encrypt the configuration.")
        data object EmptyChain : ConfError("The chain has no servers to share.")
    }

    fun resolvedPassword(password: String?): String {
        val trimmed = password?.trim().orEmpty()
        return trimmed.ifEmpty { HopConstants.APP_DISPLAY_NAME }
    }

    fun suggestedFileName(payload: Payload): String {
        val raw = when (payload) {
            is Payload.Server -> payload.profile.displayName
            is Payload.Chain -> payload.name.trim().ifEmpty { "chain" }
        }
        val safe = raw.replace('/', '-').replace(':', '-').trim().ifEmpty { "hopper" }
        return "${safe.take(40)}.$FILE_EXTENSION"
    }

    fun exportPayloadObject(payload: Payload): JSONObject {
        return when (payload) {
            is Payload.Server -> JSONObject().apply {
                put(Field.VERSION, PAYLOAD_VERSION)
                put(Field.KIND, "server")
                put(Field.SERVER, HopProfileCodec.exportJson(payload.profile))
            }
            is Payload.Chain -> {
                if (payload.hops.isEmpty()) throw ConfError.EmptyChain
                JSONObject().apply {
                    put(Field.VERSION, PAYLOAD_VERSION)
                    put(Field.KIND, "chain")
                    put(Field.NAME, payload.name)
                    put(Field.HOPS, JSONArray().apply {
                        payload.hops.forEach { put(HopProfileCodec.exportJson(it)) }
                    })
                }
            }
        }
    }

    fun exportPayloadJson(payload: Payload): String =
        exportPayloadObject(payload).toString(2)

    /** QR uses unencrypted payload. Servers stay as legacy hop-profile v2. */
    fun qrPayloadJson(payload: Payload): String = when (payload) {
        is Payload.Server -> HopQRExporter.exportJson(payload.profile)
        is Payload.Chain -> exportPayloadJson(payload)
    }

    fun parsePayloadJson(text: String): Payload {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) throw ConfError.Empty
        val json = runCatching { JSONObject(trimmed) }.getOrElse { throw ConfError.InvalidPayload }
        return parsePayloadObject(json)
    }

    fun parsePayloadObject(json: JSONObject): Payload {
        val kind = json.optString(Field.KIND, "").lowercase()
        when (kind) {
            "chain" -> {
                val name = json.optString(Field.NAME, "")
                val hopsArray = json.optJSONArray(Field.HOPS)
                    ?: throw ConfError.EmptyChain
                if (hopsArray.length() == 0) throw ConfError.EmptyChain
                val hops = buildList {
                    for (i in 0 until hopsArray.length()) {
                        val hopObj = hopsArray.optJSONObject(i) ?: throw ConfError.InvalidPayload
                        add(HopProfileCodec.parseObject(hopObj))
                    }
                }
                return Payload.Chain(name = name, hops = hops)
            }
            "server" -> {
                val serverObj = json.optJSONObject(Field.SERVER) ?: throw ConfError.InvalidPayload
                return Payload.Server(HopProfileCodec.parseObject(serverObj))
            }
            else -> return Payload.Server(HopProfileCodec.parseObject(json))
        }
    }

    fun encryptFile(payload: Payload, password: String?): ByteArray {
        val plaintext = exportPayloadJson(payload).toByteArray(Charsets.UTF_8)
        val resolved = resolvedPassword(password)
        val salt = randomBytes(SALT_LENGTH)
        val nonce = randomBytes(NONCE_LENGTH)
        val key = deriveKey(resolved, salt, PBKDF2_ITERATIONS)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
        val cipherAndTag = try {
            cipher.doFinal(plaintext)
        } catch (_: Exception) {
            throw ConfError.EncryptionFailed
        }
        val envelope = JSONObject().apply {
            put(Field.VERSION, ENVELOPE_VERSION)
            put(Field.FMT, Field.FMT_VALUE)
            put(Field.ALG, Field.ALG_VALUE)
            put(Field.KDF, Field.KDF_VALUE)
            put(Field.ITER, PBKDF2_ITERATIONS)
            put(Field.SALT, Base64.getEncoder().encodeToString(salt))
            put(Field.NONCE, Base64.getEncoder().encodeToString(nonce))
            put(Field.DATA, Base64.getEncoder().encodeToString(cipherAndTag))
        }
        return envelope.toString(2).toByteArray(Charsets.UTF_8)
    }

    fun decryptFile(data: ByteArray, password: String?): Payload {
        if (data.isEmpty()) throw ConfError.Empty
        val json = runCatching { JSONObject(String(data, Charsets.UTF_8)) }
            .getOrElse { throw ConfError.InvalidEnvelope }
        if (json.optString(Field.FMT) != Field.FMT_VALUE) throw ConfError.InvalidEnvelope

        val salt = decodeB64(json.optString(Field.SALT)) ?: throw ConfError.InvalidEnvelope
        val nonce = decodeB64(json.optString(Field.NONCE)) ?: throw ConfError.InvalidEnvelope
        val cipherAndTag = decodeB64(json.optString(Field.DATA)) ?: throw ConfError.InvalidEnvelope
        if (cipherAndTag.size <= 16) throw ConfError.InvalidEnvelope

        val iterations = json.optInt(Field.ITER, PBKDF2_ITERATIONS)
        val defaultPassword = HopConstants.APP_DISPLAY_NAME
        val provided = password?.trim().orEmpty()

        // Prefer the default password first; fall back to a user-provided password.
        openCipher(cipherAndTag, salt, nonce, iterations, defaultPassword)?.let { return it }
        if (provided.isNotEmpty() && provided != defaultPassword) {
            openCipher(cipherAndTag, salt, nonce, iterations, provided)?.let { return it }
        }
        throw ConfError.DecryptionFailed
    }

    private fun openCipher(
        cipherAndTag: ByteArray,
        salt: ByteArray,
        nonce: ByteArray,
        iterations: Int,
        password: String,
    ): Payload? {
        return try {
            val key = deriveKey(password, salt, iterations)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
            val plaintext = cipher.doFinal(cipherAndTag)
            parsePayloadJson(String(plaintext, Charsets.UTF_8))
        } catch (_: Exception) {
            null
        }
    }

    fun isHopperConfFile(data: ByteArray): Boolean {
        val json = runCatching { JSONObject(String(data, Charsets.UTF_8)) }.getOrNull() ?: return false
        return json.optString(Field.FMT) == Field.FMT_VALUE
    }

    private fun deriveKey(password: String, salt: ByteArray, iterations: Int): ByteArray {
        // UTF-8 password bytes — must match Apple CommonCrypto PBKDF2.
        val gen = PKCS5S2ParametersGenerator(SHA256Digest())
        gen.init(password.toByteArray(Charsets.UTF_8), salt, iterations)
        return (gen.generateDerivedParameters(KEY_LENGTH * 8) as KeyParameter).key
    }

    private fun randomBytes(count: Int): ByteArray {
        val bytes = ByteArray(count)
        SecureRandom().nextBytes(bytes)
        return bytes
    }

    private fun decodeB64(value: String?): ByteArray? {
        if (value.isNullOrEmpty()) return null
        return runCatching { Base64.getDecoder().decode(value) }.getOrNull()
    }
}

File: app/Shared/HopperConf.swift
import CryptoKit
import Foundation
import CommonCrypto

/// Encrypted `.hopperconf` file format + plaintext QR payload kinds.
enum HopperConf {
    static let fileExtension = "hopperconf"
    static let mimeType = "application/x-hopperconf"
    static let uti = "com.aengix.hopper.conf"
    static let envelopeVersion = 1
    static let payloadVersion = 1
    static let pbkdf2Iterations = 210_000
    static let saltLength = 16
    static let nonceLength = 12
    static let keyLength = 32

    enum Kind: String {
        case server
        case chain
    }

    enum Field {
        static let version = "v"
        static let fmt = "fmt"
        static let fmtValue = "hopperconf"
        static let alg = "alg"
        static let algValue = "aes-256-gcm"
        static let kdf = "kdf"
        static let kdfValue = "pbkdf2-sha256"
        static let iter = "iter"
        static let salt = "salt"
        static let nonce = "nonce"
        static let data = "data"
        static let kind = "kind"
        static let name = "name"
        static let server = "server"
        static let hops = "hops"
    }

    enum Payload {
        case server(HopNodeProfile)
        case chain(name: String, hops: [HopNodeProfile])
    }

    enum ConfError: LocalizedError {
        case empty
        case invalidEnvelope
        case invalidPayload
        case decryptionFailed
        case encryptionFailed
        case emptyChain

        var errorDescription: String? {
            switch self {
            case .empty: return "The file is empty."
            case .invalidEnvelope: return "Not a valid .hopperconf file."
            case .invalidPayload: return "The file contents are invalid."
            case .decryptionFailed: return "Could not decrypt — check the password."
            case .encryptionFailed: return "Could not encrypt the configuration."
            case .emptyChain: return "The chain has no servers to share."
            }
        }
    }

    /// Empty / whitespace password uses the app display name as default.
    static func resolvedPassword(_ password: String?) -> String {
        let trimmed = password?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? HopConstants.appDisplayName : trimmed
    }

    static func suggestedFileName(for payload: Payload) -> String {
        let raw: String
        switch payload {
        case .server(let profile):
            raw = profile.displayName
        case .chain(let name, _):
            raw = name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                ? "chain" : name
        }
        let safe = raw
            .replacingOccurrences(of: "/", with: "-")
            .replacingOccurrences(of: ":", with: "-")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let base = safe.isEmpty ? "hopper" : String(safe.prefix(40))
        return "\(base).\(fileExtension)"
    }

    // MARK: - Plaintext payload (QR + pre-encryption)

    static func exportPayloadDictionary(_ payload: Payload) throws -> [String: Any] {
        switch payload {
        case .server(let profile):
            return [
                Field.version: payloadVersion,
                Field.kind: Kind.server.rawValue,
                Field.server: HopProfileCodec.exportDictionary(profile),
            ]
        case .chain(let name, let hops):
            guard !hops.isEmpty else { throw ConfError.emptyChain }
            return [
                Field.version: payloadVersion,
                Field.kind: Kind.chain.rawValue,
                Field.name: name,
                Field.hops: hops.map { HopProfileCodec.exportDictionary($0) },
            ]
        }
    }

    static func exportPayloadJSON(_ payload: Payload) throws -> String {
        let dict = try exportPayloadDictionary(payload)
        let data = try JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys, .prettyPrinted])
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    /// QR uses unencrypted payload JSON. Legacy hop-profile v2 (no `kind`) is also accepted on import.
    static func qrPayloadJSON(for payload: Payload) throws -> String {
        switch payload {
        case .server(let profile):
            return HopQRExporter.exportJSON(profile)
        case .chain:
            return try exportPayloadJSON(payload)
        }
    }

    static func parsePayloadJSON(_ text: String) throws -> Payload {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw ConfError.empty }
        guard let data = trimmed.data(using: .utf8),
              let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ConfError.invalidPayload
        }
        return try parsePayloadObject(object)
    }

    static func parsePayloadObject(_ json: [String: Any]) throws -> Payload {
        let kind = stringValue(in: json, keys: [Field.kind])?.lowercased()

        if kind == Kind.chain.rawValue {
            let name = stringValue(in: json, keys: [Field.name]) ?? ""
            guard let hopsRaw = json[Field.hops] as? [[String: Any]], !hopsRaw.isEmpty else {
                throw ConfError.emptyChain
            }
            let hops = try hopsRaw.map { try HopProfileCodec.parseDictionary($0) }
            return .chain(name: name, hops: hops)
        }

        if kind == Kind.server.rawValue {
            guard let serverObj = json[Field.server] as? [String: Any] else {
                throw ConfError.invalidPayload
            }
            return .server(try HopProfileCodec.parseDictionary(serverObj))
        }

        // Legacy hop-profile v2 / deploy JSON (no kind).
        return .server(try HopProfileCodec.parseDictionary(json))
    }

    // MARK: - Encrypted file

    static func encryptFile(payload: Payload, password: String?) throws -> Data {
        let plaintext = try exportPayloadJSON(payload).data(using: .utf8) ?? Data()
        let resolved = resolvedPassword(password)
        let salt = randomBytes(saltLength)
        let nonceBytes = randomBytes(nonceLength)
        let keyData = try deriveKey(password: resolved, salt: salt)
        let key = SymmetricKey(data: keyData)
        let nonce = try AES.GCM.Nonce(data: nonceBytes)
        let sealed: AES.GCM.SealedBox
        do {
            sealed = try AES.GCM.seal(plaintext, using: key, nonce: nonce)
        } catch {
            throw ConfError.encryptionFailed
        }
        let cipherAndTag = sealed.ciphertext + sealed.tag
        let envelope: [String: Any] = [
            Field.version: envelopeVersion,
            Field.fmt: Field.fmtValue,
            Field.alg: Field.algValue,
            Field.kdf: Field.kdfValue,
            Field.iter: pbkdf2Iterations,
            Field.salt: salt.base64EncodedString(),
            Field.nonce: nonceBytes.base64EncodedString(),
            Field.data: cipherAndTag.base64EncodedString(),
        ]
        return try JSONSerialization.data(withJSONObject: envelope, options: [.sortedKeys, .prettyPrinted])
    }

    static func decryptFile(_ data: Data, password: String?) throws -> Payload {
        guard !data.isEmpty else { throw ConfError.empty }
        let object = try JSONSerialization.jsonObject(with: data)
        guard let json = object as? [String: Any] else { throw ConfError.invalidEnvelope }

        let fmt = stringValue(in: json, keys: [Field.fmt])
        guard fmt == Field.fmtValue else { throw ConfError.invalidEnvelope }

        guard let saltB64 = stringValue(in: json, keys: [Field.salt]),
              let nonceB64 = stringValue(in: json, keys: [Field.nonce]),
              let dataB64 = stringValue(in: json, keys: [Field.data]),
              let salt = Data(base64Encoded: saltB64),
              let nonceBytes = Data(base64Encoded: nonceB64),
              let cipherAndTag = Data(base64Encoded: dataB64),
              cipherAndTag.count > 16 else {
            throw ConfError.invalidEnvelope
        }

        let iterations = intValue(in: json, keys: [Field.iter]) ?? pbkdf2Iterations
        let defaultPassword = HopConstants.appDisplayName
        let provided = password?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        // Prefer the default password first; fall back to a user-provided password.
        if let payload = try? openCipher(
            cipherAndTag: cipherAndTag,
            salt: salt,
            nonceBytes: nonceBytes,
            iterations: iterations,
            password: defaultPassword
        ) {
            return payload
        }
        if !provided.isEmpty, provided != defaultPassword,
           let payload = try? openCipher(
            cipherAndTag: cipherAndTag,
            salt: salt,
            nonceBytes: nonceBytes,
            iterations: iterations,
            password: provided
           ) {
            return payload
        }
        throw ConfError.decryptionFailed
    }

    private static func openCipher(
        cipherAndTag: Data,
        salt: Data,
        nonceBytes: Data,
        iterations: Int,
        password: String
    ) throws -> Payload {
        let keyData = try deriveKey(password: password, salt: salt, iterations: iterations)
        let key = SymmetricKey(data: keyData)
        let nonce = try AES.GCM.Nonce(data: nonceBytes)
        let ciphertext = cipherAndTag.prefix(cipherAndTag.count - 16)
        let tag = cipherAndTag.suffix(16)
        let sealed = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        let plaintext = try AES.GCM.open(sealed, using: key)
        guard let text = String(data: plaintext, encoding: .utf8) else {
            throw ConfError.invalidPayload
        }
        return try parsePayloadJSON(text)
    }

    static func isHopperConfFile(_ data: Data) -> Bool {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        return stringValue(in: object, keys: [Field.fmt]) == Field.fmtValue
    }

    // MARK: - Crypto helpers

    private static func deriveKey(password: String, salt: Data, iterations: Int = pbkdf2Iterations) throws -> Data {
        var derived = Data(count: keyLength)
        let passwordBytes = Array(password.utf8)
        let status: Int32 = derived.withUnsafeMutableBytes { derivedBytes in
            salt.withUnsafeBytes { saltBytes in
                passwordBytes.withUnsafeBufferPointer { passwordPtr in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordPtr.baseAddress.map { UnsafeRawPointer($0).assumingMemoryBound(to: Int8.self) },
                        passwordBytes.count,
                        saltBytes.bindMemory(to: UInt8.self).baseAddress,
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        UInt32(iterations),
                        derivedBytes.bindMemory(to: UInt8.self).baseAddress,
                        keyLength
                    )
                }
            }
        }
        guard status == kCCSuccess else { throw ConfError.encryptionFailed }
        return derived
    }

    private static func randomBytes(_ count: Int) -> Data {
        var data = Data(count: count)
        _ = data.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, count, $0.baseAddress!) }
        return data
    }

    private static func stringValue(in dictionary: [String: Any], keys: [String]) -> String? {
        for key in keys {
            guard let value = dictionary.first(where: { $0.key.caseInsensitiveCompare(key) == .orderedSame })?.value else {
                continue
            }
            if let string = value as? String { return string }
            if let number = value as? NSNumber { return number.stringValue }
        }
        return nil
    }

    private static func intValue(in dictionary: [String: Any], keys: [String]) -> Int? {
        guard let raw = stringValue(in: dictionary, keys: keys) else { return nil }
        return Int(raw)
    }
}
